### DISCUSSION TITLE
Is there a freerdp web client?

### DISCUSSION THREAD

**Discussion by bestxpp** [FreeRDP/FreeRDP#8710] [Q&A] [answered]
Does freerdp have a web client? Or some other open source web client recommendation.
Another question, I use the same server, use freerdp and mstsc to compare, freerdp has obvious lag, mstsc is very smooth, how to make freerdp achieve the effect of mstsc

**Comment by akallabeth** [comment, marked as answer]
1. there is no FreeRDP web client
2. next best thing i can think of is [guacamole](https://guacamole.apache.org/)
3. there may be others (did not research too much on that topic)

**Reply by qxsch** [reply]
Take that one, just started today: https://github.com/qxsch/freerdp-web
works pretty good for now, looking for contributors.

- 🎬 **RDP GFX pipeline with H.264/AVC444** - Hardware-accelerated video streaming
- 🔄 **AVC444 → 4:2:0 transcoding** - Server-side FFmpeg conversion for browser compatibility
- 🖼️ **Off-main-thread rendering** - GFX Worker with OffscreenCanvas for smooth 60fps
- 🧩 **Wire format protocol** - Binary messages with typed headers (SURF, TILE, H264, etc.)
- 🎯 **Client-side GFX compositor** - Surface management, tile decoding, frame composition
- 🧮 **Progressive codec WASM decoder** - RFX Progressive tiles decoded in WebAssembly (pthreads)
- 🎨 **ClearCodec WASM decoder** - Clear codec tiles decoded in WebAssembly
- 🔊 **Low-latency audio** with AudioWorklet + SharedArrayBuffer ring buffer (~5-20ms latency)
- 🎵 Native audio streaming with Opus encoding (per-session isolation)
- ⌨️ Full keyboard support with scan code translation
- ⌨️ **Virtual on-screen keyboard** - Touch-friendly US layout with modifier support
- 🖱️ Mouse support (move, click, drag, wheel - horizontal & vertical)
- 🖱️ Server cursor support - Remote cursor updates with custom bitmaps and hotspots
- 📺 Fullscreen mode with dynamic resolution
- 📸 **Screenshot capture** - Save the current remote desktop view as an image
- 🤖 **JS-Driven automation** - Programmatic keyboard and mouse control with screenshot functionality for RPA and testing
- 🎨 **Customizable theming** - Built-in presets (dark, light, midnight, high-contrast) and shadow dom for robustness
- 📊 Latency monitoring (ping/pong)
- 🩺 Health check endpoint (`/health`)
- 🐳 Docker support with multi-stage builds
- 👥 Multi-user support (isolated RDP sessions per WebSocket connection)

**Comment by bestxpp** [comment]
What I'm using is guacamole，experience and fluency are not very good,I'm wondering if there's room for optimization。

When I don’t use gucamole, I use the xfreerdp client to compare with mstsc. xfreerdp is not as smooth as mstsc. When scrolling, there is obvious lag. Is this a problem with my settings?

**Reply by astepanov83** [reply]
@plume-cat I'm using guacamole as well, and I also noticed that it's slow, even if all the parties are located in the same network or even on the same workstation. I profiled it, and it turned out, that most of the time guacamole spends compressing screen data into png format. I replaced it on jpeg, and it became much faster, and experience became better. Also, compression and sending are done in the same receiver thread, so there is more room for optimization.

**Reply by akallabeth** [reply]
@astepanov83 I did not dig too deep in how `guacamole` does things, but sending images is quite a bad way to do it. using some video codec might be even better than your change. do you know if that was reported?

**Reply by astepanov83** [reply]
@akallabeth  I don't know, and according to their jira page, noone suggested this. Only someone sent a pull request to replace png on webp, because they too ran into the performance problem. I actually have this feature in my todo list for the app I'm working on, depending how often the performance will be an issue for the users. But no ETA yet. And there is also a lot of web-related coding, that I even cannot estimate. And I don't rely much on guacamole developers, because the fixes there are rare, and the features are added not very often.

**Reply by bestxpp** [reply]
@astepanov83 Wow, you are also using Guacamole! I tried to modify the image format before and found that the effect wasn't very noticeable. I'm planning to write my own client to replace guacd because its protocol design is too cumbersome. I intend to switch to using protobuf for directly transmitting bitmap data

**Reply by ne0YT** [reply]
@astepanov83 can you share your changes with me? i would love to try it. curently I'm using it with docker compose. maybe I can smh apply my settings to the current setup I have? :)

**Reply by qxsch** [reply]
Can you try this one? (I‘ve tried to make performant.)


https://github.com/qxsch/freerdp-web

In general it works pretty good, some tiny issues still exists.
In the future i also want to do avc444 decoding in wasm and webgpu when i find some time.

**Comment by fuqifacai** [comment]
Is that possible to call FreeRDP API or just use source codes to make it supported on web browser ?

**Comment by lazy5f** [comment]
There is [IronRDP Web Client](https://github.com/Devolutions/IronRDP/tree/master/web-client). I am not sure it fits to what you require.
